suite/tests/client-interface/drx-scattergather-bbdup.dll.c

```diff
@@ -0,0 +1,422 @@
+                       void *user_data, void *orig_analysis_data)
+{
+    /* Nop. */
+}
+
+static dr_emit_flags_t
+event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *instr,
+                      instr_t *where, bool for_trace, bool translating,
+                      void *orig_analysis_data, void *user_data)
+{
+    uint num_instrs;
+    if (user_data == NULL)
+        return DR_EMIT_DEFAULT;
+    num_instrs = *(uint *)user_data;
+    bool first = false;
+    if (drbbdup_is_first_instr(drcontext, instr, &first) != DRBBDUP_SUCCESS || !first)
+        return DR_EMIT_DEFAULT;
+    dr_insert_clean_call(drcontext, bb, where, (void *)inscount, false /* save fpstate */,
+                         1, OPND_CREATE_INT32(num_instrs));
+    /* We stress test the non-recreate path. */
+    return DR_EMIT_DEFAULT; // NOCHECK STORE_TRANSLATIONS;
+}
+
+static dr_emit_flags_t
+event_bb_analysis(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                  bool translating, void **user_data)
+{
+    uint num_sg_instrs = 0;
+    bool is_emulation = false;
+    for (instr_t *instr = instrlist_first(bb); instr != NULL;
+         instr = instr_get_next(instr)) {
+        if (instr_is_gather(instr) || instr_is_scatter(instr)) {
+            /* FIXME i#2985: some scatter/gather instructions will not get expanded in
+             * 32-bit mode.
+             */
+            IF_X64(dr_fprintf(STDERR, "Unexpected scatter or gather instruction\n"));
+        }
+        if (drmgr_is_emulation_start(instr)) {
+            emulated_instr_t emulated_instr;
+            emulated_instr.size = sizeof(emulated_instr);
+            CHECK(drmgr_get_emulated_instr_data(instr, &emulated_instr),
+                  "drmgr_get_emulated_instr_data() failed");
+            if (instr_is_gather(emulated_instr.instr) ||
+                instr_is_scatter(emulated_instr.instr))
+                num_sg_instrs++;
+            is_emulation = true;
+            continue;
+        }
+        if (drmgr_is_emulation_end(instr)) {
+            is_emulation = false;
+            continue;
+        }
+        if (is_emulation)
+            continue;
+        if (!instr_is_app(instr))
+            continue;
+    }
+    *user_data = (uint *)dr_thread_alloc(drcontext, sizeof(uint));
+    uint *num_instr_data = (uint *)*user_data;
+    *num_instr_data = num_sg_instrs;
+    return DR_EMIT_DEFAULT;
+}
+
+static dr_emit_flags_t
+event_bb_analyze_case(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                      bool translating, uintptr_t mode, void *user_data,
+                      void *orig_analysis_data, void **analysis_data)
+{
+    if (mode == INSTRU_MODE_NOP) {
+        return DR_EMIT_DEFAULT;
+    } else if (mode == INSTRU_MODE_EXPAND) {
+        return event_bb_analysis(drcontext, tag, bb, for_trace, translating,
+                                 analysis_data);
+    } else
+        DR_ASSERT(false);
+    return DR_EMIT_DEFAULT;
```

Comment:
Will it ever get to return DR_EMIT_DEFAULT? We are returning from the `if` and `else if ` clause and asserting in the `else` clause but we can only get to `return DR_EMIT_DEFAULT` if we hit the `else` but we are already asserting there.

Reply:
If the assert is debug-only (common for most assert types) then yes it will reach the return in release build, and the compiler will warn if there's no return there.  DR_ASSERT is actually enabled in release but it still seems safer to have a return after it if we move to a debug-only check.  I will move the return inside the else{}.